httpx/utils.py

```diff
@@ -212,6 +213,24 @@ def kv_format(**kwargs: typing.Any) -> str:
     return " ".join(f"{key}={value!r}" for key, value in kwargs.items())
 
 
+def get_environment_no_proxy() -> typing.List[str]:
+    """ Gets value of NO_PROXY environment value as a list"""
+    no_proxy = get_environ_lower_and_upper("NO_PROXY")
+    return no_proxy.split(",") if no_proxy else []
+
+
+def should_be_proxied(url: "URL") -> bool:
+    """ Return True if url should be proxied,
+    return False otherwise.
+    """
+    no_proxy_list = get_environment_no_proxy()
+    for exempt_url in no_proxy_list:
+        if exempt_url and url.host.endswith(exempt_url):
```

Comment:
Again following the definition from the curl manual:

> If the host name matches one of these strings, or the host is within the domain of one of these strings, transactions with that node will not be proxied. When a domain is used, it needs to start with a period. A user can specify that both www.example.com and foo.example.com should not use a proxy by setting NO_PROXY to .example.com. By including the full name you can exclude specific host names, so to make www.example.com not use a proxy but still have foo.example.com do it, set NO_PROXY to www.example.com.

Seems `url.host.endswith` is not quite enough?

Reply:
`exempt_url` can be converted to the `URL` instance. if it can, hosts match otherwise can be checked like this `if exempt_url in url.host`

Reply:
changed to using urllibs implementation, with a regular expression